src/indexer.c

```diff
@@ -153,53 +199,74 @@ RedisModuleDict * QueryIndexPredicate(RedisModuleCtx *ctx, QueryPredicate *predi
     RedisModuleDict *currentLeaf;
     RedisModuleString *index_key;
     size_t _s;
+    int nokey;
 
+    const char *key = RedisModule_StringPtrLen(predicate->label.key, &_s);
+    const char *value;
 
-    if (predicate->type == NCONTAINS || predicate->type == CONTAINS) {
-        index_key = RedisModule_CreateStringPrintf(ctx, K_PREFIX,
-                RedisModule_StringPtrLen(predicate->label.key, &_s));
+    switch (predicate->type) {
 
-    } else {
-        const char *key = RedisModule_StringPtrLen(predicate->label.key, &_s);
-        const char *value = RedisModule_StringPtrLen(predicate->label.value, &_s);
-        index_key = RedisModule_CreateStringPrintf(ctx, KV_PREFIX, key, value);
-    }
-    int nokey;
-    currentLeaf = RedisModule_DictGet(labelsIndex, index_key, &nokey);
-    if (nokey) {
-        currentLeaf = NULL;
-    } else {
-        RedisModuleDictIter *iter = RedisModule_DictIteratorStartC(currentLeaf, "^", NULL, 0);
-        RedisModuleString *currentKey;
-        while((currentKey = RedisModule_DictNext(ctx, iter, NULL)) != NULL) {
-            RedisModule_DictSet(localResult, currentKey, (void *)1);
+        case CONTAINS:
+        case NCONTAINS:
+            index_key = RedisModule_CreateStringPrintf(ctx, K_PREFIX,
+                                                       RedisModule_StringPtrLen(predicate->label.key, &_s));
+            break;
+        case NEQ:
+        case EQ:
+            value = RedisModule_StringPtrLen(predicate->label.value, &_s);
+            index_key = RedisModule_CreateStringPrintf(ctx, KV_PREFIX, key, value);
+            break;
+        default:
+            break;
+        }
+
+    if (predicate->type == FILTER_LIST) {
+
+        for (int i=0; i < predicate->label.listCount; i++){
+            const char* singleValue = RedisModule_StringPtrLen(predicate->label.valueList[i], &_s);
+            index_key = RedisModule_CreateStringPrintf(ctx, KV_PREFIX, key, singleValue);
```

Comment:
I think it's more efficient not to allocate new RedisModuleString here, just work on the two char* strings and then use 'RedisModule_DictGetC'.